src/schema.ts

```diff
@@ -247,19 +253,33 @@ function timeSummary(timeunit: TimeUnit, summary: Summary): Summary {
 /**
  * @return a new unique object based off of the old unique count and a binning scheme
  */
-function binUnique(bin, oldUnique) {
+function binUnique(bin, oldUnique, excludeInvalid: boolean) {
   const newUnique = {};
   for (var value in oldUnique) {
     let bucket: number = bin.value(Number(value)) as number;
-    if (!newUnique[bucket]) {
-      newUnique[bucket] = oldUnique[value];
-    } else {
-      newUnique[bucket] += oldUnique[value];
+    if (!excludeInvalid || (bucket !== null && bucket === bucket)) {
+      if (!newUnique[bucket]) {
+        newUnique[bucket] = oldUnique[value];
+      } else {
+        newUnique[bucket] += oldUnique[value];
+      }
     }
   }
   return newUnique;
 }
 
+// returns the number of distinct values not including values in list
+function distinctExcluding(stats: Summary, list: string[]): number {
```

Comment:
This is expensive.  Instead of running through all key in unique.  You should just have an `invalidCount` method that takes `unique` dictionary (instead of the whole summary, which is more than you need) and `list` of invalid values.  Then return numbers of values in the list and you can use this number to be subtracted from the full distinct (possibly cached).   Note that you always cache the full distinct and you can easily call this `invalidCount` to subtract every time.